Dockerfile: use default apt mirrors

crazy-max opened the pull request:
follow-up https://github.com/moby/moby/pull/44546#discussion_r1059776409

**- What I did**

Removes `BUILD_APT_MIRROR` added in https://github.com/moby/moby/pull/26375 that does not seem to be used.

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah commented:
I wonder if we should completely remove it from the Dockerfile; 

- is this useful for people behind the GFW, or not needed there?
- does "apt" /  "apt-get" have alternatives that allow passing (through env-vars?) would be nice, but don't think there is

That said; currently we unconditionally call `sed` to replace the config; perhaps we shouldn't and only do so if `APT_MIRROR` is non-empty.

thaJeztah commented on Dockerfile:
```diff
@@ -26,9 +26,6 @@ COPY --from=build-dummy /build /build
 FROM --platform=$BUILDPLATFORM ${GOLANG_IMAGE} AS base
 COPY --from=xx / /
 RUN echo 'Binary::apt::APT::Keep-Downloaded-Packages "true";' > /etc/apt/apt.conf.d/keep-cache
```
Do we know if this is still needed by the way? I see it was added in https://github.com/moby/moby/commit/82c56755ea9e6c0ce197fbdfd41d0c5d1724b50c, and I was wondering "is the default different?" (maybe it is in the base-images)

AkihiroSuda commented:
I agree with @thaJeztah , I'd prefer to retain `APT_MIRROR` in Dockerfile and Makefile.
Can be removed from Jenkinsfile though.

crazy-max committed:
Makefile: remove unused BUILD_APT_MIRROR

BUILD_APT_MIRROR added in https://github.com/moby/moby/pull/26375
is not used anymore.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: use default apt mirrors

Use default apt mirrors and also check APT_MIRROR
is set before updating mirrors.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

thaJeztah commented on Makefile:
```diff
@@ -31,7 +31,6 @@ export VALIDATE_ORIGIN_BRANCH
 # make DOCKER_LDFLAGS="-X github.com/docker/docker/daemon/graphdriver.priority=overlay2,devicemapper" dynbinary
 #
 DOCKER_ENVS := \
-	-e BUILD_APT_MIRROR \
```
Is this still needed for users that use this option? (or is it only used during `build`, in which case it needs to be added to the list of `--build-arg` (but without a value, so that it can inherit env variables with that name)

crazy-max replied:
I don't see it being used anywhere in our code base. I think it was just used to build debian packages which has been moved to docker-ce-packaging repo: https://github.com/moby/moby/pull/26375/files#diff-01ca5adb236edbbea85ba055aa48330de763cab402a67403855aa3a887ed6a07R58

neersighted commented on Dockerfile:
```diff
@@ -27,8 +27,13 @@ FROM --platform=$BUILDPLATFORM ${GOLANG_IMAGE} AS base
+RUN <<EOT
+    set -e
+    if [ -n "${APT_MIRROR}" ]; then
+        sed -ri "s/(httpredir|deb).debian.org/${APT_MIRROR:-deb.debian.org}/g" /etc/apt/sources.list
+        sed -ri "s/(security).debian.org/${APT_MIRROR:-security.debian.org}/g" /etc/apt/sources.list
```
```suggestion
        sed -ri "s/(httpredir|deb).debian.org/${APT_MIRROR}/g" /etc/apt/sources.list
        sed -ri "s/(security).debian.org/${APT_MIRROR}/g" /etc/apt/sources.list
```

These defaults don't make sense anymore as the `if` guard will prevent them from having any effect.

neersighted replied:
Also, this might be better as a one-liner at this point, e.g. `test -n "$VAR" && sed`

neersighted replied:
The latest changes here still don't address the fact that the default substitution will never take effect as `-n "$APT_MIRROR"` guarantees that `$APT_MIRROR` will be set and `${APT_MIRROR:-default}` will never use the `default` value. I think my suggested version with a single `sed` should be preferred.

crazy-max replied:
Ah I misread your review sorry :see_no_evil:

crazy-max commented:
Got many issue with the fastly CDN recently:

```
> [linux/amd64->arm64 containerd-build 2/3] RUN --mount=type=cache,sharing=locked,id=moby-containerd-aptlib,target=/var/lib/apt     --mount=type=cache,sharing=locked,id=moby-containerd-aptcache,target=/var/cache/apt         
apt-get update && xx-apt-get install -y --no-install-recommends             gcc libbtrfs-dev libsecret-1-dev:
100.8 E: Failed to fetch http://cdn-fastly.deb.debian.org/debian-security/pool/updates/main/u/util-linux/libmount-dev_2.36.1-8%2bdeb11u1_arm64.deb  Cannot initiate the connection to cdn-fastly.deb.debian.org:80 (2a04:4e42:6a::644). - connect (101: Network is unreachable) [IP: 199.232.170.132 80]
100.8 E: Failed to fetch http://cdn-fastly.deb.debian.org/debian/pool/main/p/pcre3/libpcre16-3_8.39-13_arm64.deb  Cannot initiate the connection to cdn-fastly.deb.debian.org:80 (2a04:4e42:6a::644). - connect (101: Network is unreachable) [IP: 199.232.170.132 80]
100.8 E: Failed to fetch http://cdn-fastly.deb.debian.org/debian/pool/main/p/pcre3/libpcre32-3_8.39-13_arm64.deb  Cannot initiate the connection to cdn-fastly.deb.debian.org:80 (2a04:4e42:6a::644). - connect (101: Network is unreachable) [IP: 199.232.170.132 80]
100.8 E: Failed to fetch http://cdn-fastly.deb.debian.org/debian/pool/main/g/gcc-10/libstdc%2b%2b6_10.2.1-6_arm64.deb  Cannot initiate the connection to cdn-fastly.deb.debian.org:80 (2a04:4e42:6a::644). - connect (101: Network is unreachable) [IP: 199.232.170.132 80]
100.8 E: Failed to fetch http://cdn-fastly.deb.debian.org/debian/pool/main/p/pcre3/libpcrecpp0v5_8.39-13_arm64.deb  Cannot initiate the connection to cdn-fastly.deb.debian.org:80 (2a04:4e42:6a::644). - connect (101: Network is unreachable) [IP: 199.232.170.132 80]
100.8 E: Failed to fetch http://cdn-fastly.deb.debian.org/debian/pool/main/p/pcre3/libpcre3-dev_8.39-13_arm64.deb  Cannot initiate the connection to cdn-fastly.deb.debian.org:80 (2a04:4e42:6a::644). - connect (101: Network is unreachable) [IP: 199.232.170.132 80]
100.8 E: Failed to fetch http://cdn-fastly.deb.debian.org/debian-security/pool/updates/main/z/zlib/zlib1g-dev_1.2.11.dfsg-2%2bdeb11u2_arm64.deb  Cannot initiate the connection to cdn-fastly.deb.debian.org:80 (2a04:4e42:6a::644). - connect (101: Network is unreachable) [IP: 199.232.170.132 80]
100.8 E: Failed to fetch http://cdn-fastly.deb.debian.org/debian/pool/main/g/glib2.0/libglib2.0-dev_2.66.8-1_arm64.deb  Cannot initiate the connection to cdn-fastly.deb.debian.org:80 (2a04:4e42:6a::644). - connect (101: Network is unreachable) [IP: 199.232.170.132 80]
100.8 E: Failed to fetch http://cdn-fastly.deb.debian.org/debian/pool/main/libs/libsecret/libsecret-1-dev_0.20.4-2_arm64.deb  Cannot initiate the connection to cdn-fastly.deb.debian.org:80 (2a04:4e42:6a::644). - connect (101: Network is unreachable) [IP: 199.232.170.132 80]
100.8 E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
------
Dockerfile:208
--------------------
 207 |     ARG TARGETPLATFORM
 208 | >>> RUN --mount=type=cache,sharing=locked,id=moby-containerd-aptlib,target=/var/lib/apt \
 209 | >>>     --mount=type=cache,sharing=locked,id=moby-containerd-aptcache,target=/var/cache/apt \
 210 | >>>         apt-get update && xx-apt-get install -y --no-install-recommends \
 211 | >>>             gcc libbtrfs-dev libsecret-1-dev
 212 |     ARG DOCKER_STATIC
--------------------
ERROR: failed to solve: process "/bin/sh -c apt-get update && xx-apt-get install -y --no-install-recommends             gcc libbtrfs-dev libsecret-1-dev" did not complete successfully: exit code: 100
```

Would appreciate to have this one in :pray:

neersighted commented:
I think this is the best middle ground; unconditionally using Fastly isn't great. We should also consider not overriding `security.debian.org` since many mirrors do not carry the `debian-security` repo, but that can be a follow-up.

cpuguy83 reviewed (approved):
Reading through the changes this looks fine.
